In [1]:
from IPython.display import display, clear_output, HTML
import time

def countdown_timer(minutes):
    total_seconds = minutes * 60
    for seconds in range(total_seconds, 0, -1):
        mins, secs = divmod(seconds, 60)
        time_str = f"{mins:02}'{secs:02}''"
        clear_output(wait=True)
        # HTML with styling
        display(HTML(f'<div style="font-size: 24px; color: blue; font-weight: bold;">Time remaining: {time_str}</div>'))
        time.sleep(1)
    clear_output(wait=True)
    # Final message with different styling
    display(HTML('<div style="font-size: 24px; color: green; font-weight: bold;">Time\'s up!</div>'))

<div style="text-align: center;">
    <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/f/f8/Python_logo_and_wordmark.svg/972px-Python_logo_and_wordmark.svg.png?20210516005643" alt="The Python logo" style="width: 60%; max-width: 500px;">
</div>

<div class="alert alert-block alert-warning">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

# Functions 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

# Funktionen
</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
        
The code we've written so far lacks reusability. 

To enhance this, it's useful to organize code into named blocks, or functions, that can be called repeatedly, potentially with different inputs. 

Writing functions not only improves code portability and reduces repetition but also simplifies complex tasks into smaller, more manageable pieces. 

This leads to code that is easier to understand, maintain, and debug.
        
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
        
Dem Code, den wir bisher geschrieben haben, mangelt es an Wiederverwendbarkeit. 

Um dies zu verbessern, ist es nützlich, den Code in benannte Blöcke oder Funktionen zu gliedern, die wiederholt aufgerufen werden können, möglicherweise mit unterschiedlichen Eingaben. 
Das Schreiben von Funktionen verbessert nicht nur die Übertragbarkeit des Codes und reduziert Wiederholungen, sondern vereinfacht auch komplexe Aufgaben in kleinere, besser handhabbare Teile. 

Dies führt zu Code, der leichter zu verstehen, zu warten und zu debuggen ist.
        
</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
        
A function is defined with a `def` statement which gives it a name and specifies any <i>arguments</i> that it takes in brackets, followed by a colon. 

The body of the function is indented by 4 spaces. 
        
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
        
Eine Funktion wird mit einer `def`-Anweisung definiert, die ihr einen Namen gibt und alle <i>Argumente</i> angibt, die sie in Klammern, gefolgt von einem Doppelpunkt, annimmt.

Der Körper der Funktion wird um 4 Leerzeichen eingerückt.
        
</div>
</div>

```python
def <function_name>(<[arguments]>):
    <body>
    return <value>
```

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
        
When called, the function’s statements are executed; if the `return` keyword is encountered at any point, the specified values are returned to the caller:
        
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
        
Wenn die Funktion aufgerufen wird, werden die Anweisungen der Funktion ausgeführt; wenn das Schlüsselwort `return` an irgendeiner Stelle auftaucht, werden die angegebenen Werte an den Aufrufer zurückgegeben:
        
</div>
</div>

In [5]:
# Defining the function:
def hello_world():
    return("Hello, World!")


# Calling the function:
hello_world()

'Hello, World!'

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
        
The function `hello_world` does not accept any arguments, but often we want functions to take input so they can operate on different values. To achieve this, you simply need to specify the argument inside the parentheses when defining the function.
        
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
        
The function `hello_world` does not accept any arguments, but often we want functions to take input so they can operate on different values. To achieve this, you simply need to specify the argument inside the parentheses when defining the function.
        
</div>
</div>

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### Exercise

Write a function that takes a number `x` as an argument and returns its cube.

Test the function on a number of your choice.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### Übung

Schreiben Sie eine Funktion, die eine Zahl `x` als Argument nimmt und ihren Kubus zurückgibt.

Testen Sie die Funktion mit einer Zahl Ihrer Wahl.
</div>
</div>

In [ ]:
#countdown
countdown_timer(5)

In [7]:
def cube(x):
    x_cube = x**3
    return x_cube

cube(3)

27

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
    
####  Passing multiple arguments 
        
The function `cube` takes a single argument, `x` (the number to be cubed), and returns a single value (its cube). 
Of course, you can pass multiple arguments to a function. 
Simply list the arguments inside the parentheses, separated by commas, allowing the function to work with more than one input.
        
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
    
####  Mehrere Argumente übergeben 
        
The function `cube` takes a single argument, `x` (the number to be cubed), and returns a single value (its cube). 
Of course, you can pass multiple arguments to a function. 
Simply list the arguments inside the parentheses, separated by commas, allowing the function to work with more than one input.
        
</div>
</div>

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### Exercise

Write a function that takes two numbers `r` and `h` as arguments and returns the volume of a cylinder, $V = \pi r^2 h$.

Test the function on two numbers of your choice.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### Übung

Schreiben Sie eine Funktion, die zwei Zahlen `r` und `h` als Argumente annimmt und das Volumen eines Zylinders, $V = \pi r^2 h$, zurückgibt.

Testen Sie die Funktion mit zwei Zahlen Ihrer Wahl.
</div>
</div>

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### Exercise

Write a function that takes the lengths of the two shorter sides of a right triangle as input and returns the length of the hypotenuse.

Test the function on two numbers of your choice.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### Übung

Write a function that takes the lengths of the two shorter sides of a right triangle as input and returns the length of the hypotenuse.

Testen Sie die Funktion mit zwei Zahlen Ihrer Wahl.
</div>
</div>

In [ ]:
#countdown
countdown_timer(10)

In [2]:
import math 

def cylinder(r,h):
    V = math.pi * r**2 * h
    return V

cylinder(2,5)

62.83185307179586

In [17]:
def pythagoras(a,b):
    c = math.sqrt(a**2 + b**2)
    return c

pythagoras(3,4)

5.0

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
        
<i>In Python, you should write `return c` rather than `return(c)`.

Both will work in practice, but `return c` is the preferred syntax according to the Python style guide (PEP 8). 

This is because `return` is not a function, so parentheses are unnecessary. 

The parentheses in `return(c)` might confuse readers into thinking `return` is a function, which it is not.</i>
        
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
        
<i>In Python sollte man `return c` und nicht `return(c)` schreiben.

Beide funktionieren in der Praxis, aber `return c` ist die bevorzugte Syntax nach dem Python Style Guide (PEP 8).

Das liegt daran, dass `return` keine Funktion ist, so dass Klammern unnötig sind.

Die Klammern in `return(c)` könnten den Leser zu der Annahme verleiten, dass `return` eine Funktion ist, was nicht der Fall ist.</i>
        
</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

####  Returning multiple values 
        
And, of course, to return two or more values from a function, separate them by commas in the `return` statement.
        
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
  
####  Mehrere Werte zurückgeben 
        
Wenn Sie zwei oder mehr Werte aus einer Funktion zurückgeben möchten, trennen Sie diese durch Kommas in der `return`-Anweisung.
        
</div>
</div>

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
        #### Exercise

Recall the quadratic formula and define a function that takes the polynomial coefficients `a`, `b`, and `c` as arguments, then returns the roots of the equation.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
        #### Übung

Erinnern Sie sich an die quadratische Formel und definieren Sie eine Funktion, die die Polynomkoeffizienten `a`, `b` und `c` als Argumente nimmt und dann die Wurzeln der Gleichung zurückgibt.
</div>
</div>
<div style="text-align: center; margin-top: 20px;">

$$
ax^2 + bx + c = 0  \implies x_{1,2} = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}
$$
</div>

In [ ]:
#countdown
countdown_timer(5)

In [3]:
def roots(a, b, c):
    d = b**2 - 4*a*c 
    r1 = (-b + math.sqrt(d)) / 2 / a 
    r2 = (-b - math.sqrt(d)) / 2 / a 
    return r1, r2

roots(1,3,-10)

(2.0, -5.0)

In [6]:
type(roots(1,3,-10))

tuple

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
        
When you use `return` followed by multiple values separated by commas (e.g., `return r1, r2`), Python internally groups (<i>packs</i>) these values into a **tuple**.

This means that the multiple values are stored in a single tuple object, which can then be unpacked later if needed.

You can also unpack the tuple directly into variables:
        
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
        
Wenn Sie `return` gefolgt von mehreren durch Kommas getrennten Werten verwenden (z.B. `return r1, r2`), gruppiert (<i>packt</i>) Python intern diese Werte in ein **Tupel**.

Das bedeutet, dass mehrere Werte in einem einzigen Tupel-Objekt gespeichert werden, das dann später bei Bedarf wieder entpackt werden kann.

Sie können das Tupel auch direkt in Variablen entpacken:
        
</div>
</div>

In [21]:
r1, r2 = roots(1,3,-10)

print(r1)
print(r2)

2.0
-5.0


<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
        
<i>You technically can use tuple indexing to access individual elements of a tuple, but it’s not as efficient or readable when trying to "unpack" the values. 

Tuple unpacking and tuple indexing are two different approaches, and they serve different purposes.

Tuple unpacking is the preferred and cleaner method when you know the exact number of elements in the tuple. 

It allows you to directly assign values to multiple variables in a single step, making the code more readable.</i>
        
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">
        
<i>Technisch gesehen kann man die Tupel-Indizierung verwenden, um auf einzelne Elemente eines Tupels zuzugreifen, aber es ist nicht so effizient oder lesbar, wenn man versucht, die Werte "auszupacken". 

Tupel-Entpacken und Tupel-Indizierung sind zwei verschiedene Ansätze, die unterschiedlichen Zwecken dienen.

Das Entpacken von Tupeln ist die bevorzugte und sauberere Methode, wenn Sie die genaue Anzahl der Elemente in einem Tupel kennen. 

Sie ermöglicht es Ihnen, mehreren Variablen in einem einzigen Schritt direkt Werte zuzuweisen, was den Code lesbarer macht.</i>
        
</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

## Keywords Arguments
        
So far, we have used <i>positional arguments</i> when calling functions, where the values are assigned based on the order in which they are passed. </br>

For example, `roots(1, 3, -10)` calls the function with `a=1, b=3, c=-10`. 

Alternatively, you can pass values in any order by using <i>keyword arguments</i>, which explicitly assign values to specific parameters by name.
        
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Schlüsselwort-Argumente
        
Bislang haben wir beim Aufruf von Funktionen <i>Positionsargumente</i> verwendet, bei denen die Werte in der Reihenfolge der Übergabe zugewiesen werden.

Zum Beispiel ruft `roots(1, 3, -10)` die Funktion mit `a=1, b=3, c=-10` auf. 

Alternativ können Sie Werte in beliebiger Reihenfolge übergeben, indem Sie <i>Schlüsselwortargumente</i> verwenden, die bestimmten Parametern explizit Werte namentlich zuweisen.
        
</div>
</div>

In [13]:
roots(1,3,-10) == roots(a=1, b=3, c=-10)

True

In [14]:
roots(1,3,-10) == roots(b=3, c=-10, a=1)

True

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

## Default Arguments 
        
Sometimes it is helpful to make an argument to a function optional; that is, the function code will use a default value for that argument if one is not provided when the function is called. 

This can be done in the function definition:
        
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Standardargumente
        
Manchmal ist es hilfreich, ein Argument für eine Funktion optional zu machen, d.h. der Funktionscode verwendet einen Standardwert für dieses Argument, wenn es beim Aufruf der Funktion nicht angegeben wird. 

Dies kann in der Funktionsdefinition erfolgen:
        
</div>
</div>

In [16]:
def roots(a=1, b=3, c=-10):
    d = b**2 - 4*a*c 
    r1 = (-b + math.sqrt(d)) / 2 / a 
    r2 = (-b - math.sqrt(d)) / 2 / a 
    return r1, r2

roots()     # we can call the function with no arguments, because default values have been provided in the definition

(2.0, -5.0)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

Now, the function can be called with no arguments.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Nun kann die Funktion ohne Argumente aufgerufen werden.
</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">
    
##  Variable scope 

Every function we've written so far has included a return statement, even though it's optional: Functions can execute their tasks without one. 

However, to understand why we frequently use return statements, it's important to explore how Python handles variables within functions.

Let’s revisit the function we defined earlier, `cube(x)`, but this time, we'll use a `print` statement instead of a `return` statement:
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

## Variablenbereich 

Jede Funktion, die wir bisher geschrieben haben, enthält eine Rückgabeanweisung, auch wenn sie optional ist: Funktionen können ihre Aufgaben auch ohne eine solche ausführen. 

Um jedoch zu verstehen, warum wir häufig Return-Anweisungen verwenden, ist es wichtig zu verstehen, wie Python mit Variablen innerhalb von Funktionen umgeht.

Wenden wir uns noch einmal der Funktion zu, die wir vorhin definiert haben, `cube(x)`, aber dieses Mal verwenden wir eine `print`-Anweisung anstelle einer `return`-Anweisung:
</div>
</div>

In [9]:
def cube(x):
    x_cube = x**3
    print(x_cube)

cube(3)

27


<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

**In this version, the function will print the result of cubing x but won't return a value that can be used elsewhere in your code.**

This means that while you can see the output, you won't be able to store or manipulate the result of the function.

In fact, if we attempt to access the variable *x_cube* by calling it, we'll encounter an error:
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

**In dieser Version gibt die Funktion das Ergebnis der Kubikbildung von x aus, liefert aber keinen Wert, der an anderer Stelle in Ihrem Code verwendet werden kann.**

Das bedeutet, dass Sie zwar die Ausgabe sehen können, aber nicht in der Lage sind, das Ergebnis der Funktion zu speichern oder zu manipulieren.

Wenn wir nämlich versuchen, auf die Variable *x_cube* zuzugreifen, indem wir sie aufrufen, wird ein Fehler auftreten:
</div>
</div>

In [10]:
x_cube

NameError: name 'x_cube' is not defined

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

In Python, <i>variable scope</i> refers to where a variable is accessible within the code. 

Variables can be classified as either <i>global</i> or <i>local</i> based on where they are defined:
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

In Python bezieht sich der <i>Variablenbereich</i> darauf, wo eine Variable innerhalb des Codes zugänglich ist.

Variablen können entweder als <i>global</i> oder <i>lokal</i> klassifiziert werden, je nachdem, wo sie definiert sind:
</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

-  **Global** variables are defined outside of functions and are accessible throughout the entire program, including inside functions. 

-  **Local** variables are defined within a function and can only be accessed inside that function. Once the function ends, local variables are discarded, and trying to access them outside the function results in an error.

</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

-  **Globale** Variablen werden außerhalb von Funktionen definiert und sind im gesamten Programm zugänglich, auch innerhalb von Funktionen.

-  **Lokale** Variablen werden innerhalb einer Funktion definiert und können nur innerhalb dieser Funktion aufgerufen werden. Sobald die Funktion endet, werden lokale Variablen verworfen, und der Versuch, auf sie außerhalb der Funktion zuzugreifen, führt zu einem Fehler.

</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

To use the value of a local variable later in your code, you need to <i>capture the returned value</i> from the function into a new variable: 

-  Return the value from the function using a `return` statement.

-  Store the returned value in a new variable outside the function.


**While the variable itself remains local to the function, the value it holds can be returned and then assigned to a different variable outside the function.**
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Um den Wert einer lokalen Variablen später in Ihrem Code zu verwenden, <i>müssen Sie den von der Funktion zurückgegebenen Wert in eine neue Variable einfangen</i>: 

-  Geben Sie den Wert der Funktion mit einer `return`-Anweisung zurück.

-  Speichern Sie den zurückgegebenen Wert in einer neuen Variablen außerhalb der Funktion.

**Während die Variable selbst lokal in der Funktion bleibt, kann der Wert, den sie enthält, zurückgegeben und dann einer anderen Variablen außerhalb der Funktion zugewiesen werden.**
</div>
</div>

In [11]:
def cube(x):
    x_cube = x**3       # 'x_cube' is a local variable
    return x_cube

cubed_value = cube(3)   # capturing the returned value in a new variable 

In [13]:
cubed_value             # now you can use 'cubed_value' later in your code

27

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

# Exercises
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

# Übungen
</div>
</div>

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### 1

The code below attempts to calculate the balance of a savings account at the end of each year for a maximum of four years.  The account starts with a balance of 100 EUR, and the annual interest rate is 0.72%.  However, the code is not functioning as expected. Identify the issue(s) and offer a corrected version.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### 1

Der folgende Code versucht, den Saldo eines Sparkontos am Ende eines jeden Jahres für maximal vier Jahre zu berechnen.  Das Konto beginnt mit einem Saldo von 100 EUR, und der jährliche Zinssatz beträgt 0,72 %.  Der Code funktioniert jedoch nicht wie erwartet. Ermitteln Sie die Fehler und bieten Sie eine korrigierte Version an.
</div>
</div>

In [14]:
# The following code does not work as intended - identify the problem, and correct it in the cell below

balance = 100

def add_interest(balance, rate):
    balance += balance * rate/100
    
for year in range(1, 5):
    add_interest(balance, 0.72)
    print ('Balance after year', year, '=', balance)

Balance after year 1 = 100
Balance after year 2 = 100
Balance after year 3 = 100
Balance after year 4 = 100


In [ ]:
countdown_timer(10)

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

####  Solution 

The code does not work as expected because of how variable scope works in Python, specifically regarding the `balance` variable inside the function.

The original code doesn't work as expected because the function is modifying a local copy of the `balance` variable, not the global `balance`. To fix this, you need to return the updated value from the function and reassign it outside the function.

1. **Local vs. Global Variables**:
   - The `balance` variable is defined globally (outside of any function) with an initial value of `100`.
   - Inside the `add_interest` function, Python treats `balance` as a **local variable** because you are trying to modify it (`balance += balance * rate/100`).
   - This creates a **new local variable** `balance` within the function, which is separate from the global `balance` variable.
   - As a result, the global `balance` variable remains unchanged outside the function, and the local `balance` only exists within the function.

2. **No Return Statement**:
   - The function `balance` modifies the **local** `balance` variable, but since it doesn't return anything, the changes made inside the function are not reflected outside the function.
   - To fix this, you need to return the updated balance from the function and reassign it to the global `balance`.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### Lösung 

Der Code funktioniert nicht wie erwartet, weil der Variablenbereich in Python so funktioniert, insbesondere in Bezug auf die `balance`-Variable innerhalb der Funktion.
Der ursprüngliche Code funktioniert nicht wie erwartet, weil die Funktion eine lokale Kopie der Variable `balance` ändert, nicht die globale Variable `balance`. Um dies zu beheben, müssen Sie den aktualisierten Wert aus der Funktion zurückgeben und ihn außerhalb der Funktion neu zuweisen.

1. **Lokale vs. globale Variablen**:
   - Die Variable `balance` ist global (außerhalb einer Funktion) mit einem Anfangswert von "100" definiert.
   - Innerhalb der Funktion `add_interest` behandelt Python die Variable "Saldo" als **lokale Variable**, da Sie versuchen, sie zu verändern (`balance += balance * rate/100`).
   - Dadurch wird innerhalb der Funktion eine **neue lokale Variable** `balance` erzeugt, die von der globalen Variable `Guthaben` getrennt ist.
   - Dadurch bleibt die globale Variable `balance` außerhalb der Funktion unverändert, und die lokale Variable `balance` existiert nur innerhalb der Funktion.

2. **Keine Rückgabeanweisung**:
   - Die Funktion `add_interest` ändert die **lokale** Variable `balance`, aber da sie nichts zurückgibt, werden die innerhalb der Funktion vorgenommenen Änderungen außerhalb der Funktion nicht berücksichtigt.
   - Um dies zu beheben, müssen Sie den aktualisierten Saldo aus der Funktion zurückgeben und ihn erneut der globalen `balance`-Variable zuweisen.
</div>
</div>

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

You can modify the function to return the updated balance and then capture the returned value in the main code loop.

In this solution:
- The function returns the updated `balance`, which is then reassigned to the global `balance` variable after each year.
- This way, the global `balance` is updated correctly, and the output reflects the change.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Sie können die Funktion so ändern, dass sie den aktualisierten Saldo zurückgibt und dann den zurückgegebenen Wert in der Hauptcodeschleife erfassen.

In dieser Lösung:
- Die Funktion gibt den aktualisierten `balance` zurück, der dann nach jedem Jahr der globalen `balance`-Variablen neu zugewiesen wird.
- Auf diese Weise wird der globale `balance` korrekt aktualisiert, und die Ausgabe spiegelt die Änderung wider.
</div>
</div>

In [17]:
balance = 100

def add_interest(balance, rate):
    balance += balance * rate / 100
    return balance  # Return the updated balance

# Now capture the returned value in the loop
for year in range(1, 5):
    balance = add_interest(balance, 0.72)  # Update balance with the returned value
    print('Balance after year', year, '=', balance)

Balance after year 1 = 100.72
Balance after year 2 = 101.445184
Balance after year 3 = 102.1755893248
Balance after year 4 = 102.91125356793856


<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### 2

Write a simple program to numerically differentiate mathematical functions using the **first-principles definition** of a derivative:
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### 2

Schreiben Sie ein einfaches Programm zur numerischen Differenzierung mathematischer Funktionen unter Verwendung der **Erstprinzipien-Definition** einer Ableitung:
</div>
</div>
<div style="text-align: center; margin-top: 20px;">

$$
f'(x) = \lim_{h \to 0} \frac{f(x+h) - f(x)}{h}
$$
</div>
<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

Here, $f(x)$ is the original function, $f'(x)$ is its derivative, and $h$ is the step size. 

This equation shows that to compute a derivative numerically, we only need the definition of the function $f(x)$ and a value for the step size—numerical differentiation simply requires that we can define the function!

As a basic example, let's calculate the derivative of $f(x) = x^4$ at $x = 1.3$. 

The analytical solution, $f'(x) = 4x^3$, gives a result of 8.788. 

1. Define the mathematical function $f(x)=x^{4}$ using python.

2. Using the first principles definition of a derivative, write a second function, `derivative()`, containing two parameters: the stepsize, $h$ (which we shall give a default value of 0.01), and the value of $x$ we are using to evaluate the derivative.

**Useful tips:**

- Functions can be called from within other functions, and their output used as normal.

- Both `x` and `h` are numbers, therefore `x+h` is also a number, and can be passed to a function as an argument.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

$f(x)$ ist die Funktion, $f'(x)$ ist ihre Ableitung und $h$ ist die Schrittweite. 

Diese Gleichung zeigt, dass wir für die numerische Berechnung einer Ableitung nur die Definition der Funktion $f(x)$ und einen Wert für die Schrittweite benötigen - die numerische Differenzierung erfordert lediglich, dass wir die Funktion definieren können!

Berechnen wir als einfaches Beispiel die Ableitung von $f(x) = x^4$ bei $x = 1.3$. 

Die analytische Lösung, $f'(x) = 4x^3$, liefert ein Ergebnis von 8.788.

1. Definieren Sie die mathematische Funktion $f(x)=x^{4}$ mit Python.

2. Schreiben Sie unter Verwendung der Definition einer Ableitung nach den ersten Prinzipien eine zweite Funktion, `derivative()`, die zwei Parameter enthält: die Schrittweite $h$ (die wir mit 0.01 vorgeben) und den Wert von $x$, den wir zur Auswertung der Ableitung verwenden.

**Nützliche Tipps:**

- Funktionen können aus anderen Funktionen heraus aufgerufen werden, und ihre Ausgabe wird wie üblich verwendet.

- Sowohl `x` als auch `h` sind Zahlen, daher ist `x+h` auch eine Zahl und kann als Argument an eine Funktion übergeben werden.
</div>
</div>

In [20]:
def f(x):
    y = x**4
    return y

f(8)

4096

In [21]:
def derivative(x, h=0.01):
    f_prime = (f(x+h) - f(x)) / h
    return f_prime

derivative(1.3, 0.01)

8.889921000000012

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

Your `derivative()` function should have returned a value of approximately 8.89 for the derivative of $x^4$ evaluated at $x=1.3$. 

While this is close to the exact value of 8.788, can we improve the accuracy?

The first-principles definition of a derivative assumes the limit as $h$ approaches zero. 

This implies that by decreasing $h$ (getting closer to $h = 0$), our numerical approximation should become more precise. 
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

Ihre Funktion `Ableitung()` sollte für die Ableitung von $x^4$ nach $x=1.3$ einen Wert von etwa 8.89 liefern. 

Dies liegt zwar nahe am exakten Wert von 8,788, aber können wir die Genauigkeit verbessern?

Die Definition einer Ableitung nach den ersten Grundsätzen geht davon aus, dass der Grenzwert bei $h$ gegen Null geht. 

Das bedeutet, dass unsere numerische Näherung mit abnehmendem $h$ (näher an $h = 0$) präziser werden sollte.
</div>
</div>

<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### 3

Using your `derivative()` function, repeat the calculation above with step sizes of 
$$ h = 1, 0.1, 0.01, 0.001, 0.0001, 0.00001.$$ 

Try to incorporate a loop in your code to automate the function calls, instead of manually invoking the function for each step size.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### 3

Wiederholen Sie die obige Berechnung mit der Funktion `derivative()` mit Schrittweiten von
$$ h = 1, 0.1, 0.01, 0.001, 0.0001, 0.00001.$$ 

Versuchen Sie, eine Schleife in Ihren Code einzubauen, um die Funktionsaufrufe zu automatisieren, anstatt die Funktion für jede Schrittweite manuell aufzurufen.
</div>
</div>

In [22]:
step_sizes = [1, 0.1, 0.01, 0.001, 0.0001, 0.00001]

for h in step_sizes:
    print(derivative(1.3, h))

25.12799999999999
9.85500000000001
8.889921000000012
8.798145200998864
8.789014051999722
8.788101400547532


<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### 4

Write a program that can compute the derivative with respect to $x$ for any mathematical function of the form:
$f(x) = ax^p$.

Test your program by evaluating the derivative of $f(x) = 4x^3$ at $x = 3$. 

If your code is working correctly, it should return a value close to 108 (the exact value may vary slightly depending on the step size chosen).
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### 4

Schreiben Sie ein Programm, das die Ableitung nach $x$ für eine beliebige mathematische Funktion der Form berechnen kann:
$f(x) = ax^p$.

Testen Sie Ihr Programm, indem Sie die Ableitung von $f(x) = 4x^3$ nach $x = 3$ auswerten. 

Wenn Ihr Programm korrekt arbeitet, sollte es einen Wert nahe 108 zurückgeben (der genaue Wert kann je nach der gewählten Schrittweite leicht variieren).
</div>
</div>

In [23]:
def f(x, a, p):
    y = 4*x**p
    return y

def derivative(x, a, p, h):
    value = (f(x+h, a, p) - f(x, a, p)) / h
    print(value)
    
derivative(x=3, a=4, p=3, h=1.e-6)

108.00003602184916


<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### 5

Write a Python function to calculate the heat capacity of sodium and diamond according to the Einstein model given below.

$\Theta_E$ is 192 K for sodium and 1450 K for diamond.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### 5

Schreiben Sie eine Python-Funktion zur Berechnung der Wärmekapazität von Natrium und Diamant nach dem unten angegebenen Einstein-Modell.

$\Theta_E$ ist 192 K für Natrium und 1450 K für Diamant.
</div>
</div>
<div style="text-align: center; margin-top: 20px;">

$$
C_{V,m} = 3R \left( \frac{\Theta_E}{T} \right)^2 \frac{\exp\left( \frac{\Theta_E}{T} \right)}{\left[ \exp\left( \frac{\Theta_E}{T} \right) - 1 \right]^2}
$$
</div>
<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

**Hints**:

- import $R$ from `scipy.constants`,

- define a variable $\Theta_E / T$ for your convenience,

- remember to import all relevant modules.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

**Hinweise**:

- importieren Sie $R$ aus `scipy.constants`,

- definieren Sie eine Variable $\Theta_E / T$ für Ihre Bequemlichkeit,

- Denken Sie daran, alle relevanten Module zu importieren.
</div>
</div>

In [27]:
from scipy.constants import R
import math

def C_V(T, Theta_E):
    x = Theta_E / T
    return 3 * R * x**2 * math.exp(x) / (math.exp(x) - 1)**2

print(f"Na: {C_V(300, 192)}")
print(f"diamond: {C_V(300, 1450)}")

Na: 24.109144168075208
diamond: 4.713033453464986


<div class="alert alert-block alert-light">

<div style="display: flex; justify-content: space-between;">
<div style="width: 48%; line-height: 1.3;">

#### 6

For the following equations of state, define Python functions returning the pressure of a gas, $p$, given its volume $V$, amount $n$, and temperature $T$.

- **Ideal gas equation of state**: $$ p = \frac{nRT}{V},$$
- **Van der Waals equation of state**: $$ p = \frac{nRT}{V-nb} - \frac{n^2a}{V^2},$$
- **Dieterici equation of state**: $$ p = \frac{nRT \exp \left( -\frac{nc}{RTV} \right)}{v-nd}.$$

Use the following parameters:

- T = 273.15 K;
- V = 20 L;
- a = 3.640 $L^2$ bar $mol^{-2}$, b = 0.04267 L $mol^{-1}$;
- c = 4.692 $L^2$ bar $mol^{-2}$, d = 0.04639 L $mol^{-1}$;
- n = 1 mol.

Repeat the calculation for n = 0.1 mol.

Print summaries using string formatting.
</div>
<div style="width: 48%; line-height: 1.3;color: grey;">

#### 6

Definieren Sie für die folgenden Zustandsgleichungen Python-Funktionen, die den Druck eines Gases $p$ wiedergeben, wenn das Volumen $V$, die Menge $n$ und die Temperatur $T$ gegeben sind.

- **Ideale Gaszustandsgleichung**: $$ p = \frac{nRT}{V},$$
- **Van-der-Waals-Zustandsgleichung**: $$ p = \frac{nRT}{V-nb} - \frac{n^2a}{V^2},$$
- **Dieterici-Zustandsgleichung**: $$ p = \frac{nRT \exp \left( -\frac{nc}{RTV} \right)}{v-nd}.$$

Verwenden Sie die folgenden Parameter:

- T = 273.15 K;
- V = 20 L;
- a = 3.640 $L^2$ bar $mol^{-2}$, b = 0.04267 L $mol^{-1}$;
- c = 4.692 $L^2$ bar $mol^{-2}$, d = 0.04639 L $mol^{-1}$;
- n = 1 mol.

Wiederholen Sie die Berechnung für n = 0.1 mol.

Zusammenfassungen unter Verwendung der String-Formatierung drucken.
</div>
</div>